runtime/src/accounts_db.rs

```diff
@@ -2170,34 +2171,48 @@ impl AccountsDb {
             .expect("Cluster type must be set at initialization")
     }
 
-    /// Reclaim older states of accounts older than max_clean_root_inclusive for AccountsDb bloat mitigation
+    /// Reclaim older states of accounts older than max_clean_root_inclusive for AccountsDb bloat mitigation.
+    /// Any accounts which are removed from the accounts index are returned in PubkeysRemovedFromAccountsIndex.
+    /// These should NOT be unref'd later from the accounts index.
     fn clean_accounts_older_than_root(
         &self,
         purges: Vec<Pubkey>,
         max_clean_root_inclusive: Option<Slot>,
         ancient_account_cleans: &AtomicU64,
-    ) -> ReclaimResult {
+    ) -> (ReclaimResult, PubkeysRemovedFromAccountsIndex) {
+        let pubkeys_removed_from_accounts_index = HashSet::default();
         if purges.is_empty() {
-            return ReclaimResult::default();
+            return (
+                ReclaimResult::default(),
+                pubkeys_removed_from_accounts_index,
+            );
         }
         // This number isn't carefully chosen; just guessed randomly such that
         // the hot loop will be the order of ~Xms.
         const INDEX_CLEAN_BULK_COUNT: usize = 4096;
 
         let one_epoch_old = self.get_accounts_hash_complete_one_epoch_old();
+        let pubkeys_removed_from_accounts_index = Mutex::new(pubkeys_removed_from_accounts_index);
 
         let mut clean_rooted = Measure::start("clean_old_root-ms");
         let reclaim_vecs = purges
             .par_chunks(INDEX_CLEAN_BULK_COUNT)
             .filter_map(|pubkeys: &[Pubkey]| {
                 let mut reclaims = Vec::new();
                 for pubkey in pubkeys {
-                    self.accounts_index.clean_rooted_entries(
+                    let removed_from_index = self.accounts_index.clean_rooted_entries(
                         pubkey,
                         &mut reclaims,
                         max_clean_root_inclusive,
                     );
+                    if removed_from_index {
+                        pubkeys_removed_from_accounts_index
+                            .lock()
+                            .unwrap()
+                            .insert(*pubkey);
```

Comment:
this is one place where we have to remember that we previously removed this pubkey from the accounts index already.